zone/draw_teams.php

```diff
@@ -176,25 +187,181 @@ private static function permute_match_sizes(array $match_sizes): array
+        $odd_indices = array_fill(0, $number_matches, 0);
+
+        $j = 1;
+        for ($i = 0; $i < $number_matches - 1; $i += 2) {
+            $even_indices[$i] = $j;
+            $even_indices[$i + 1] = $j;
+            $j++;
+        }
+
+        $j = 1;
+        for ($i = 1; $i < $number_matches - 1; $i += 2) {
+            $odd_indices[$i] = $j;
+            $odd_indices[$i + 1] = $j;
+            $j++;
+        }
+
+        return [$even_indices, $odd_indices];
+    }
+
+    public function permute_edge_players(
+        entity\match_players_list $upper_players,
+        entity\match_players_list $lower_players,
+        int $number
+    ): array {
+        if ($number == 0) {
+            return [[$upper_players, $lower_players]];
+        }
+
+        if ($number == 1) {
+            $default = [$upper_players, $lower_players];
+            $upper_players = clone $upper_players;
+            $lower_players = clone $lower_players;
+
+            $upper_player = $upper_players->pop();
+            $lower_player = $lower_players->shift();
+            $upper_players->add($lower_player);
+            $lower_players->unshift($upper_player);
+            $switched = [$upper_players, $lower_players];
+
+            return [$default, $switched];
+        }
+
+        if ($number == 2) {
+            $switched_player_permutations = [];
+
+            $upper_players = clone $upper_players;
+            $lower_players = clone $lower_players;
+
+            $players_in_question = [
+                $upper_players->pop(),
+                $upper_players->pop(),
+                $lower_players->shift(),
+                $lower_players->shift()
+            ];
+
+            foreach (self::$permute_players_def as $permutations) {
+                $temp_upper_players = clone $upper_players;
+                $temp_lower_players = clone $lower_players;
+
+                $upper_indices = $permutations[0];
+                $lower_indices = $permutations[1];
+                $temp_upper_players->add(
+                    $players_in_question[$upper_indices[0]],
+                    $players_in_question[$upper_indices[1]]
+                );
+                $temp_lower_players->unshift(
+                    $players_in_question[$lower_indices[0]],
+                    $players_in_question[$lower_indices[1]]
+                );
+
+                $switched_player_permutations[] = [$temp_upper_players, $temp_lower_players];
+            }
+
+            return $switched_player_permutations;
+        }
+
+        throw new \InvalidArgumentException('$number may only be one of 0, 1 or 2');
+    }
+
     public function make_match(
         entity\match_players_list $player_list,
         float $factor
     ): entity\draw_match {
-        [$match] = $this->make_matches($player_list, $factor);
+        [$match] = $this->make_matches($player_list, $factor, 0, 0);
         return $match;
     }
 
+    public function permute_player_lists(
```

Comment:
unit test adden plz. 

das ist sicher die fkt die so unlesbar ist. aber wenn du die testest wird es vll wenigstens einfacher, die schoen zu machen, weil man dann gleich sieht, ob man was zerhaut :D

Reply:
Puhh... da muss ich mir mal was überlegen. Die erzeugt wahnsinnig viel Output, da hat man mit 1 Testcase schon paar hundert Zeilen

Reply:
vielleicht hilft es, die fkt zu unterteilen in testbare abschnitte? ^^ aber oki meinetwegen passt das so, wir muessen hier ja nicht auf max test abdeckung bestehen :D